osu.Game.Rulesets.Osu/Difficulty/Preprocessing/L2NormCorrection.cs

```diff
@@ -0,0 +1,210 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Linq;
+using MathNet.Numerics;
+using MathNet.Numerics.Interpolation;
+
+namespace osu.Game.Rulesets.Osu.Difficulty.Preprocessing
+{
+    public class L2NormCorrection
+    {
+        public L2NormCorrection(double[] distance, double[] x_offset, double[] y_offset, double[] offset, double[] scale)
+        {
+            this.x_offset = LinearSpline.InterpolateSorted(distance, x_offset);
+            this.y_offset = LinearSpline.InterpolateSorted(distance, y_offset);
+            this.offset = LinearSpline.InterpolateSorted(distance, offset);
+            this.scale = LinearSpline.InterpolateSorted(distance, scale);
+        }
+
+        public double Evaluate(double distance, double x, double y)
+        {
+            x -= x_offset.Interpolate(distance);
+            y -= y_offset.Interpolate(distance);
+            double z = offset.Interpolate(distance);
+            double c = scale.Interpolate(distance);
+            return c * Math.Sqrt(x * x + y * y + z);
+        }
+
+        private LinearSpline x_offset, y_offset, offset, scale;
+    }
+
+    public class MultiL2NormCorrection
+    {
+        public MultiL2NormCorrection(L2NormCorrection[] components, double[] distance, double[] offset, double[] scale)
+        {
+            this.offset = LinearSpline.InterpolateSorted(distance, offset);
+            this.scale = LinearSpline.InterpolateSorted(distance, scale);
+            this.components = components;
+        }
+
+        public double Evaluate(double distance, double x, double y)
+        {
+            double result = components.Select(component => component.Evaluate(distance, x, y)).Sum();
+            result += offset.Interpolate(distance);
+            return SpecialFunctions.Logistic(result) * scale.Interpolate(distance);
+        }
+
+        private L2NormCorrection[] components;
+        private LinearSpline offset, scale;
+
+        private static readonly double[] flow0distances = { 0.0, 1.0, 1.35, 1.7, 2.3, 3 };
+        public static readonly MultiL2NormCorrection FLOW_0 = new MultiL2NormCorrection(
```

Comment:
also i feel like there should be a brief explanation of flow/snap naming as well